### ISSUE TITLE
Return types of [Weakly]Incrementable op++ and op++(int) are overconstrained

### ISSUE THREAD

**Issue by ericniebler** [ericniebler/stl2#137]
Many output iterators have an op++ that do nothing and simply return a `const` ref to *this. (This would be perfectly acceptable for the ostream and insert iterators, for instance.) Since WeaklyIncrementable is looking for an exact match to `I&`, such an iterator does not satisfy `Iterator`, although it probably should.

We can define a `OneOf` concept like:

```c++
template <class T, class...Us>
concept bool OneOf() {
  return Same<T, Us>() || ...;
}
```

and define [Weakly]Incrementable like this:

```c++
template <class I>
concept bool WeaklyIncrementable() {
  return Semiregular<I>() &&
    requires {
        typename difference_type_t<I>;
        { ++i } -> OneOf<I&, const I&>;
        i++;
    }
}

template <class I>
concept bool Incrementable() {
  return WeaklyIncrementable<I>() &&
    EqualityComparable<I> &&
    requires {
        { i++ } -> OneOf<I, I&, const I&>;
    }
}
```

@CaseyCarter, @asutton, thoughts?

**Comment by CaseyCarter** [comment]
I'm not enthused: the disjunctions in the generated constraint expressions would be deleterious to compile times. C++ 14 requires `++o` to return a mutable reference to `o`, so `WeaklyIncrementable`'s requirement on `++i` doesn't exclude any conforming STL1 output iterators. C++14 `InputIterator` also requires `++i` to return a mutable reference, although it strangely doesn't require that reference to refer to `i` rendering the return value effectively useless (this is probably a defect). I don't see any benefit to returning a `const&` instead of a mutable reference from the pre-increment operator, so I don't think we are ruling out any interesting future iterator models with the current requirement.

I have less strong feelings about post-increment, given that it's already a mess. I'd prefer for `Incrementable` to have `int`-like behavior absent other strong considerations. Given that `WeaklyIncrementable` already doesn't care about `i++`, and `Incrementable` isn't required by anything below `ForwardIterator`, `Incrementable`'s requirements don't affect any hypothetical models of output or input iterators.

**Comment by asutton** [comment]
>
> I'm not enthused: the disjunctions in the generated constraint expressions
> would be deleterious to compile times.
>

There should be no intent to overload on them (the ordering makes no
sense), so making OneOf a constexpr function would be reasonable. Not every
predicate needs to be a concept.

++i should always return a mutable reference. I consider ++(++i) to be
valid expression for all iterators (assuming reachability).

post-increment is a mess. It would be better if we didn't require it at
all.

C++ 14 requires ++o to return a mutable reference to o, so

**Comment by ericniebler** [comment]
> ++i should always return a mutable reference. I consider ++(++i) to be
valid expression for all iterators (assuming reachability).

I think both you and Casey have missed my point. The op++ for ostream_iterator does *nothing* and so can both return a const& and itself be const, so that ++(++i) compiles just fine. In addition, op* can be be both const and return a const & to \*this, so that `*(++(++i)) = 42` compiles just fine.

```c++
struct ostream_iterator {
    std::ostream *sout;
    template<class T>
    const ostream_iterator& operator=(T const &t) const { *sout << t; return *this;}
    const ostream_iterator& operator*() const { return *this;}
    const ostream_iterator& operator++() const { return *this;}
    const ostream_iterator& operator++(int) const { return *this;}
    /// ...
};
```

There's nothing wrong with the above iterator -- it works just fine in practice -- except that it doesn't satisfy `WeaklyIncrementable` so it's not an iterator. That's annoying to people like me who are in the good habit of putting const everywhere we can.

**Comment by asutton** [comment]
I don't know... pick a direction. Weaken the concept or invalidate the
type. My preference is the latter.

It could be argued that the consts on pre-increment and = are inappropriate
for that type. Even though *sout is not a subobject, the intent is clearly
to modify *something* observable.


Andrew Sutton

**Comment by CaseyCarter** [comment]
> It could be argued that the consts on pre-increment and = are inappropriate
for that type. Even though *sout is not a subobject, the intent is clearly
to modify *something* observable.

I would say this implementation of `ostream_iterator` violates [res.on.data.races]/3:

> A C++ standard library function shall not directly or indirectly modify objects (1.10) accessible by threads other than the current thread unless the objects are accessed directly or indirectly via the function’s non-const arguments, including `this`.

`this` is `const` in `operator=`, yet it modifies `*this->sout` which may be accessible to other threads. Even though [res.on.data.races] does not apply to non-standard-library functions - say if you'd called this `my_ostream_iterator` - I consider this a violation of `const`-correctness. I would take issue with the design in a code review. For example, it doesn't satisfy `Assignable<ostream_iterator&, int>()`, which I think could be problematic for generic code.

More generally, if `++i` might return a `const&` then we lose the capability to pass it to a function that takes a mutable reference. I just don't see enough benefit here to justify the cost.

**Comment by ericniebler** [comment]
> I would say this implementation of ostream_iterator violates [res.on.data.races]/3:
>
> > A C++ standard library function shall not directly or indirectly modify objects (1.10) accessible by threads other than the current thread unless the objects are accessed directly or indirectly via the function’s non-const arguments, including `this`.

Wouldn't this also be violated by any mutation of an element accessed through an iterator with a `const`-qualified `operator*`?

> For example, it doesn't satisfy `Assignable<ostream_iterator&, int>()`, which I think could be problematic for generic code.

That's a red herring. No generic code would ever test such a condition. You must dereference an iterator before you write through it. `Writable<ostream_iterator&, int>` would be satisfied, and that's all that matters.

> More generally, if ++i might return a const& then we lose the capability to pass it to a function that takes a mutable reference.

`advance(++i, n)` would work just fine. I don't see the problem.

**Comment by CaseyCarter** [comment]
> Wouldn't this also be violated by any mutation of an element accessed through an iterator with a const-qualified operator*?

The `operator*` is non-modifying. To mutate an element you must apply some other modifying operation to the result of `operator*`.

> No generic code would ever test such a condition.

I was thinking of a naive user requiring `Assignable<reference_t<I>, int>` instead of `Writable<I, int>`, which in any case would work just fine. You're right here ;)

> advance(++i, n) would work just fine. I don't see the problem.

`advance` is specified to accept a mutable reference, it wouldn't compile.

**Comment by ericniebler** [comment]
> Wouldn't this also be violated by any mutation of an element accessed through an iterator with a const-qualified operator*?
>> The `operator*` is non-modifying. To mutate an element you must apply some other modifying operation to the result of `operator*`.

Precisely. The `operator*` is a const-qualified function that gives you **access** to an object that is then mutated. Read the text of [res.on.data.races]/3 again and tell me it doesn't apply to exactly this case:

> A C++ standard library function shall not directly or indirectly modify objects (1.10) accessible by threads other than the current thread **unless the objects are accessed directly or indirectly via the function’s non-const arguments**

If you interpret my `ostream_iterator` as violating this passage, we have to also conclude that `std::sort` violates it also when passed iterators with const-qualified op*, since the objects are being access through a const-qualified member function.

> `advance` is specified to accept a mutable reference, it wouldn't compile.

Sure it would. `template<class I> void advance(I&,int)` ... `I` would be deduced as `const ostream_iterator`, which can be incremented. (A finer point is whether that type satisfies WeaklyIncrementable. which is exactly the topic under discussion.)

Let me try again. `ostream_iterator` is (somewhat cleverly) using itself as a proxy reference type. You guys seem to be confusing the const-ness of the proxy as applying to the const-ness of the (virtual, in this case) object being proxied. Proxies don't work that way. Their mutability derives from the mutability of the proxied object because you assign *through* the proxy.

**Comment by CaseyCarter** [comment]
I stand corrected, the call to `advance(++i)` compiles because the `const` assignment operator is unconstrained. The call is in fact ill-formed NDR since `const ostream_iterator` doesn't satisfy `Copyable`. You need to add a `const ostream_iterator& operator=(const ostream_iterator&) const` and make `sout` mutable in order for `const ostream_iterator` to satisfy `Semiregular`.

**Comment by CaseyCarter** [comment]
> Their mutability derives from the mutability of the proxied object because you assign through the proxy.

Any operation that can break substitution in equality-preserving expressions is a modification. I.e., if `a` equals `b` and `(f(a), a)` doesn't equal `b` then `f` must be a modification. We've forced reference-semantic entities into our value-semantic framework by defining that two references are equal if the entities they denote are equal. As a general rule, we've decided that expressions that operate on `const` entities are not modifications, and expressions that operate on non-`const` entities may be modifications.

I think for consistency with that model an operation that can break substitutability of a reference entity in equality-preserving expressions must be non-`const`. The fact that a reference entity may also happen to be an object, and that an operation can modify its referent without physically altering the bits of that object's representation is immaterial. One could argue that the value denoted by a pure output iterator is unobservable and hence *no* operation on its reference type can modify any observable value, but to me it makes more sense for the declarations of operations on an output iterator's reference type to be consistent with those of a readable iterator's reference type. Assignment to the reference type alters the notional value that it denotes even if that value is unobservable.

Yes, this same argument indicates that `reference_wrapper`'s member `operator T&() const` should be replaced with `operator const T&() const` and `operator T&()`.

> If you interpret my ostream_iterator as violating this passage, we have to also conclude that std::sort violates it also when passed iterators with const-qualified op*, since the objects are being access through a const-qualified member function.

If you said `std::all_of` here, I would agree that it may not mutate the values of the denoted objects. The mutating algorithms are each individually specified to mutate the objects denoted by their arguments, however, which takes precedence over the general rule. You could argue that the assignment operator of `ostream_iterator`'s reference type (which in this case is `const ostream_iterator`) is also specified to mutate the notional object it denotes, and I would have to agree that [res.on.data.races] doesn't apply. So I'll concede the point.

**Comment by ericniebler** [comment]
> Yes, this same argument indicates that `reference_wrapper`'s member `operator T&() const` should be replaced with `operator const T&() const` and `operator T&()`.

I need to think more about this when I'm not rushed trying to get updated drafts ready for the post-meeting mailing but this seems ... odd. `reference_wrapper` is intended to be as reference-like as possible. The change you're suggesting breaks my mental model of what a reference is. I think of a reference as a pointer that is implicitly dereferenced when it is accessed (read or written). Like pointers, top-level consts do not affect the const-ness of the object referenced. It's so nonsensical that we're not even allowed to const-qualify reference types. Why should a top-level const on `reference_wrapper` change the const-ness of the referent? I'll go back and reread what you wrote and sit with it. It may have implications for the proxy iterators paper.

**Comment by CaseyCarter** [comment]
> Yes, this same argument indicates that `reference_wrapper`'s member `operator T&() const` should be replaced with `operator const T&() const` and `operator T&()`.

Not exactly. `operator T&() const` is in fact non-modifying. It returns a modifiable lvalue, but neither the `reference_wrapper` object nor its referent are modified. To actually *modify* the referent one would need to apply another expression which *is* a modification to the returned lvalue.

That is at least notionally not the case in the `ostream_iterator` example. The proxy reference has a `const` member that modifies its (again purely notional) referent. I could analogously define a new member for `reference_wrapper`:
```c++
void foo() const {
  *ptr = T{};
}
```
which is well-formed according to the core language since it doesn't change the bits of the `reference_wrapper` object. That operation is *definitely* performing a modification to something observable.

**Comment by ericniebler** [comment]
Superseded by #232.
